### DISCUSSION TITLE
大佬们,能不能把3.x版本的wiki恢复回来呀?4.x的改了配置方法,真抓瞎...

### DISCUSSION THREAD

**Discussion by Tmalll** [IrineSistiana/mosdns#468] [Q&A]
之前3.x写了个境内外分流又去广告的配置, 想改改又找不到3.x的wiki了...
想换4.x版本, 可4的wiki写的没3详细, 很多逻辑不知道咋搞...
或者哪位大佬能给个4.x的境内外分流又去广告的配置模板哈...

**Comment by NoyAcg2022NoyAcg2022s** [comment]
仅供参考 瞎写
log:
  level: error
  file: "/file"
data_providers:
  - tag: site
    file: "/geosite.dat"
    auto_reload: true
  - tag: ip
    file: "/geoip.dat"
    auto_reload: true
  - tag: gfwip
    file: "/gfwip"
    auto_reload: true
  - tag: lili
    file: "/lili"
    auto_reload: true
  - tag: lilis
    file: "/lilis"
    auto_reload: true
  - tag: ococ
    file: "/ococ"
    auto_reload: true
  - tag: ococs
    file: "/ococs"
    auto_reload: true
  - tag: aa
    file: "/aa"
    auto_reload: true
  - tag: aas
    file: "/aas"
    auto_reload: true
  - tag: auto
    file: "/auto"
    auto_reload: true
plugins:
  - tag: cn
    type: fast_forward
    args:
      upstream:
        - addr: "tls://120.53.53.53:853"
          trusted: true
          idle_timeout: 40
          enable_pipeline: true
          max_conns: 1
          insecure_skip_verify: false
  - tag: cns
    type: fast_forward
    args:
      upstream:
        - addr: "tls://120.53.53.53:853"
          trusted: true
          idle_timeout: 40
          enable_pipeline: true
          max_conns: 1
          insecure_skip_verify: false
  - tag: cnE
    type: fast_forward
    args:
      upstream:
        - addr: "tls://120.53.53.53:853"
          trusted: true
          idle_timeout: 40
          enable_pipeline: true
          max_conns: 1
          insecure_skip_verify: false
  - tag: gle
    type: fast_forward
    args:
      upstream:
        - addr: "tls://8.8.4.4:853"
          trusted: true
          idle_timeout: 10
          enable_pipeline: true
          max_conns: 1
          insecure_skip_verify: false
  - tag: gles
    type: fast_forward
    args:
      upstream:
        - addr: "tls://8.8.4.4:853"
          trusted: true
          idle_timeout: 10
          enable_pipeline: true
          max_conns: 1
          insecure_skip_verify: false
  - tag: gleE
    type: fast_forward
    args:
      upstream:
        - addr: "tls://8.8.4.4:853"
          trusted: true
          idle_timeout: 10
          enable_pipeline: true
          max_conns: 1
          insecure_skip_verify: false
  - tag: isCN
    type: query_matcher
    args:
      domain:
        - "provider:ococ"
        - "provider:ococs"
        - "provider:site:cn"
  - tag: isAds
    type: query_matcher
    args:
      domain:
        - "provider:lili"
        - "provider:lilis"
        - "provider:site:category-ads-all,win-spy,win-extra"
  - tag: isA
    type: query_matcher
    args:
      domain:
        - "provider:ococ"
        - "provider:ococs"
        - "provider:aa"
  - tag: isNO
    type: query_matcher
    args:
      domain:
        - "provider:site:geolocation-!cn"
  - tag: isIP
    type: response_matcher
    args:
      ip:
        - "provider:ip:cn"
  - tag: isgfwip
    type: response_matcher
    args:
      ip:
        - "provider:gfwip"
  - tag: redirect
    type: redirect
    args:
      rule:
        - "aod.itunes.apple.com hls.itunes.apple.com.mwcname.com"
        - "speech.platform.bing.com cn.bing.com"
  - tag: zaizais
    type: response_matcher
    args:
      rcode: [2]
  - tag: is12
    type: query_matcher
    args:
      qtype: [12]
      domain:
        - "provider:aas"
  - tag: is65
    type: query_matcher
    args:
      qtype: [65]
  - tag: cache
    type: cache
    args:
      size: 102400
      lazy_cache_ttl: 25200
  - tag: ttl
    type: ttl
    args:
      minimal_ttl: 300
      maximum_ttl: 25200
  - tag: ttls
    type: ttl
    args:
      minimal_ttl: 30
      maximum_ttl: 25200
  - tag: auto
    type: hosts
    args:
      hosts:
        - "provider:auto"
  - tag: udpme
    type: sequence
    args:
      exec:
        - _misc_optm
        - cache
        - auto
        - redirect
        - _no_ecs
        - if: is65
          exec:
            - _drop_response
            - _return
        - if: "(isAds) && (! isA) || (is12)"
          exec:
            - _new_nxdomain_response
            - _return
        - if: isCN
          exec:
            - load_balance:
                - - cn
                - - cns
                - - cnE
            - _return
        - if: isNO
          exec:
            - _prefer_ipv4
            - load_balance:
                - - gle
                - - gles
                - - gleE
            - _return
        - primary:
            - load_balance:
                - - cn
                - - cns
                - - cnE
            - ttl
            - if: "(! isIP) && [_response_valid_answer] || (isgfwip)"
              exec:
                - _drop_response
          secondary:
            - _prefer_ipv4
            - load_balance:
                - - gle
                - - gles
                - - gleE
            - ttls
            - if: zaizais
              exec:
                - load_balance:
                    - - cn
                    - - cns
                    - - cnE
                - ttls
          fast_fallback: 150
servers:
  - exec: udpme
    timeout: 5
    listeners:
      - protocol: udp
        addr: "127.0.0.1:53"
      - protocol: udp
        addr: "[::1]:53"
api:
  http: "127.0.0.1:3000"

求这几个排列顺序 
        - _misc_optm
        - cache
        - auto
        - redirect
        - _no_ecs

**Comment by Tmalll** [comment]
> 仅供参考 瞎写 log: level: error file: "/file" data_providers:
> 
> * tag: site
>   file: "/geosite.dat"
>   auto_reload: true
> * tag: ip
>   file: "/geoip.dat"
>   auto_reload: true
> * tag: gfwip
>   file: "/gfwip"
>   auto_reload: true
> * tag: lili
>   file: "/lili"
>   auto_reload: true
> * tag: lilis
>   file: "/lilis"
>   auto_reload: true
> * tag: ococ
>   file: "/ococ"
>   auto_reload: true
> * tag: ococs
>   file: "/ococs"
>   auto_reload: true
> * tag: aa
>   file: "/aa"
>   auto_reload: true
> * tag: aas
>   file: "/aas"
>   auto_reload: true
> * tag: auto
>   file: "/auto"
>   auto_reload: true
>   plugins:
> * tag: cn
>   type: fast_forward
>   args:
>   upstream:
>   - addr: "tls://120.53.53.53:853"
>   trusted: true
>   idle_timeout: 40
>   enable_pipeline: true
>   max_conns: 1
>   insecure_skip_verify: false
> * tag: cns
>   type: fast_forward
>   args:
>   upstream:
>   - addr: "tls://120.53.53.53:853"
>   trusted: true
>   idle_timeout: 40
>   enable_pipeline: true
>   max_conns: 1
>   insecure_skip_verify: false
> * tag: cnE
>   type: fast_forward
>   args:
>   upstream:
>   - addr: "tls://120.53.53.53:853"
>   trusted: true
>   idle_timeout: 40
>   enable_pipeline: true
>   max_conns: 1
>   insecure_skip_verify: false
> * tag: gle
>   type: fast_forward
>   args:
>   upstream:
>   - addr: "tls://8.8.4.4:853"
>   trusted: true
>   idle_timeout: 10
>   enable_pipeline: true
>   max_conns: 1
>   insecure_skip_verify: false
> * tag: gles
>   type: fast_forward
>   args:
>   upstream:
>   - addr: "tls://8.8.4.4:853"
>   trusted: true
>   idle_timeout: 10
>   enable_pipeline: true
>   max_conns: 1
>   insecure_skip_verify: false
> * tag: gleE
>   type: fast_forward
>   args:
>   upstream:
>   - addr: "tls://8.8.4.4:853"
>   trusted: true
>   idle_timeout: 10
>   enable_pipeline: true
>   max_conns: 1
>   insecure_skip_verify: false
> * tag: isCN
>   type: query_matcher
>   args:
>   domain:
>   - "provider:ococ"
>   - "provider:ococs"
>   - "provider:site:cn"
> * tag: isAds
>   type: query_matcher
>   args:
>   domain:
>   - "provider:lili"
>   - "provider:lilis"
>   - "provider:site:category-ads-all,win-spy,win-extra"
> * tag: isA
>   type: query_matcher
>   args:
>   domain:
>   - "provider:ococ"
>   - "provider:ococs"
>   - "provider:aa"
> * tag: isNO
>   type: query_matcher
>   args:
>   domain:
>   - "provider:site:geolocation-!cn"
> * tag: isIP
>   type: response_matcher
>   args:
>   ip:
>   - "provider:ip:cn"
> * tag: isgfwip
>   type: response_matcher
>   args:
>   ip:
>   - "provider:gfwip"
> * tag: redirect
>   type: redirect
>   args:
>   rule:
>   - "aod.itunes.apple.com hls.itunes.apple.com.mwcname.com"
>   - "speech.platform.bing.com cn.bing.com"
> * tag: zaizais
>   type: response_matcher
>   args:
>   rcode: [2]
> * tag: is12
>   type: query_matcher
>   args:
>   qtype: [12]
>   domain:
>   - "provider:aas"
> * tag: is65
>   type: query_matcher
>   args:
>   qtype: [65]
> * tag: cache
>   type: cache
>   args:
>   size: 102400
>   lazy_cache_ttl: 25200
> * tag: ttl
>   type: ttl
>   args:
>   minimal_ttl: 300
>   maximum_ttl: 25200
> * tag: ttls
>   type: ttl
>   args:
>   minimal_ttl: 30
>   maximum_ttl: 25200
> * tag: auto
>   type: hosts
>   args:
>   hosts:
>   - "provider:auto"
> * tag: udpme
>   type: sequence
>   args:
>   exec:
>   - _misc_optm
>   - cache
>   - auto
>   - redirect
>   - _no_ecs
>   - if: is65
>   exec:
>   - _drop_response
>   - _return
>   - if: "(isAds) && (! isA) || (is12)"
>   exec:
>   - _new_nxdomain_response
>   - _return
>   - if: isCN
>   exec:
>   - load_balance:
>   - - cn
>   - - cns
>   - - cnE
>   - _return
>   - if: isNO
>   exec:
>   - _prefer_ipv4
>   - load_balance:
>   - - gle
>   - - gles
>   - - gleE
>   - _return
>   - primary:
>   - load_balance:
>   - - cn
>   - - cns
>   - - cnE
>   - ttl
>   - if: "(! isIP) && [_response_valid_answer] || (isgfwip)"
>   exec:
>   - _drop_response
>   secondary:
>   - _prefer_ipv4
>   - load_balance:
>   - - gle
>   - - gles
>   - - gleE
>   - ttls
>   - if: zaizais
>   exec:
>   - load_balance:
>   - - cn
>   - - cns
>   - - cnE
>   - ttls
>   fast_fallback: 150
>   servers:
> * exec: udpme
>   timeout: 5
>   listeners:
>   
>   * protocol: udp
>     addr: "127.0.0.1:53"
>   * protocol: udp
>     addr: "[::1]:53"
>     api:
>     http: "127.0.0.1:3000"
> 
> 求这几个排列顺序 - _misc_optm - cache - auto - redirect - _no_ecs

下面这是我的3.x的配置, 照着之前wiki里大佬的配置改的.

log:
  level: debug   # 日志级别。可选 `debug` `info` `warn` `error`。默认 `info`。
  file: "./Log_Main_All.log"

plugin:
  - tag: main_server
    type: server
    args:
      entry:
        - _no_ecs
        - _prefer_ipv4
        - _single_flight
        - main_sequence
        - modify_ttl
      timeout: 10
      server:
        - protocol: udp
          addr: "0.0.0.0:53"

        - protocol: tcp
          addr: "0.0.0.0:53153"

        - protocol: 'doh'           # 必需。服务器协议。可以是 `udp`, `tcp`, `dot/tls`, `doh/https`, `http`。
          addr: '0.0.0.0:53443'     # 必需。服务器监听地址。省略 IP 会监听所有地址(包括 IPv6)。
          idle_timeout: 10          # 连接复用空连接超时时间。单位: 秒。默认: 10。
          cert: '/usr/local/etc/xray/Acme_ADGC.crt'   # `dot`, `doh` 必需。PEM 格式的证书文件。
          key: '/usr/local/etc/xray/Acme_ADGC.key'    # `dot`, `doh` 必需。PEM 格式的密钥文件。
          url_path: '/dns-query'    # url 路径。适用于 `doh`, `http`。如果为空，则任意路径的请求都能被处理。
          get_user_ip_from_header: 'X-Forwarded-For' # 尝试从 HTTP 头获取用户 IP。适用于 `doh`, `http`。如果获取不到依旧使用来源 IP。

        - protocol: 'dot'           # 必需。服务器协议。可以是 `udp`, `tcp`, `dot/tls`, `doh/https`, `http`。
          addr: '0.0.0.0:53853'     # 必需。服务器监听地址。省略 IP 会监听所有地址(包括 IPv6)。
          idle_timeout: 10          # 连接复用空连接超时时间。单位: 秒。默认: 10。
          cert: '/usr/local/etc/xray/Acme_ADGC.crt'   # `dot`, `doh` 必需。PEM 格式的证书文件。
          key: '/usr/local/etc/xray/Acme_ADGC.key'    # `dot`, `doh` 必需。PEM 格式的密钥文件。

  - tag: main_sequence
    log: 
      level: 'debug' # 日志级别。可选 `debug` `info` `warn` `error`。默认 `info`。
      file: './Log_main_sequence.log'
    type: sequence
    args:
      exec:
        - if: #屏蔽广告
            - query_is_ad_domain
          exec:
            - _block_with_nxdomain
            - _return

        - lazy_cache # 此处禁用缓存

        - if: #强制境内
            - Force_forward_local
          exec:
            - forward_local
            - _return

        - if: #强制境外
            - Force_forward_remote
          exec:
            - forward_remote
            - _return

        - if: #境内
            - query_is_local_domain
            - "!_query_is_common"
          exec:
            - forward_local
            - _return

        - if: #境外
            - query_is_non_local_domain
          exec:
            - _prefer_ipv4
            - forward_remote
            - _return
        - primary:
            - forward_local
            - if:
                - "!response_has_local_ip"
              exec:
                - _drop_response
          secondary:
            - _prefer_ipv4
            - forward_remote
          fast_fallback: 300
          always_standby: true 
  
  # 缓存设置
  - tag: lazy_cache
    type: cache
    args:
      size: 51200
      lazy_cache_ttl: 300
      lazy_cache_reply_ttl: 300

  # 修改应答 ttl
  - tag: 'modify_ttl'
    type: 'ttl'
    args:
      minimal_ttl: 300
      maximum_ttl: 300

  - tag: forward_local
    log: 
      level: 'debug' # 日志级别。可选 `debug` `info` `warn` `error`。默认 `info`。
      file: './Log_forward_local.log'
    type: fast_forward
    args:
      timeout: 10
      upstream:
        # 阿里DNS
        - addr: 'https://223.5.5.5/dns-query'
          trusted: true
          enable_pipeline: true
        - addr: 'https://223.6.6.6/dns-query'
          trusted: true
          enable_pipeline: true
        #
        #
        #
                 
        # 腾讯DNS
        - addr: 'https://doh.pub/dns-query'
          dial_addr: '1.12.12.12:443'
          enable_pipeline: true
        - addr: 'https://doh.pub/dns-query'
          dial_addr: '120.53.53.53:443'
          enable_pipeline: true
        - addr: 'https://doh.pub/dns-query'
          dial_addr: '162.14.21.178:443'
          enable_pipeline: true
        - addr: 'https://doh.pub/dns-query'
          dial_addr: '162.14.21.56:443'
          enable_pipeline: true
        #
        #
        #
          
        # 360 DNS
        - addr: 'https://doh.360.cn/dns-query'
          dial_addr: '112.65.69.15:443'
          enable_pipeline: true
        - addr: 'https://doh.360.cn/dns-query'
          dial_addr: '123.6.48.18:443'
          enable_pipeline: true
        #
        #
        #
  
  - tag: forward_remote
    log: 
      level: 'debug' # 日志级别。可选 `debug` `info` `warn` `error`。默认 `info`。
      file: './Log_forward_remote.log'
    type: fast_forward
    args:
      timeout: 10
      upstream:
        # Cloudflare_DOH
        - addr: 'https://1.1.1.1/dns-query'
          socks5: '192.168.1.20:10800'
          trusted: true
          enable_pipeline: true
        - addr: 'https://1.0.0.1/dns-query'
          socks5: '192.168.1.20:10800'
          trusted: true
          enable_pipeline: true
        #
        #
        #
        
        # Google_DOH
        - addr: 'https://dns.google/dns-query'
          dial_addr: '8.8.8.8:443'
          socks5: '192.168.1.20:10800'
          trusted: true
          enable_pipeline: true
        - addr: 'https://dns.google/dns-query'
          dial_addr: '8.8.4.4:443'
          socks5: '192.168.1.20:10800'
          trusted: true
          enable_pipeline: true
        #
        #
        #

        # iQDNS DNS https://iqdns.xyz/all.html
        # 
        # 北京腾讯云 25ms
        - addr: 'tls://b.iqiq.io:853'
          enable_pipeline: true
        # 
        # 上海电信 40ms
        - addr: 'tls://x.passcloud.xyz'
          enable_pipeline: true
        #  
        # 广州腾讯云 40ms
        - addr: 'tls://c.passcloud.xyz'
          enable_pipeline: true
        #  
        #  香港腾讯云a 50ms
        - addr: 'https://a.passcloud.xyz/dns-query'
          enable_pipeline: true
        #  
        #  香港腾讯云i 50ms
        - addr: 'https://i.passcloud.xyz/dns-query'
          enable_pipeline: true
        #
        #
        #

        # IIJ DNS 70ms
        - addr: 'https://public.dns.iij.jp/dns-query'
          dial_addr: '103.2.57.6:443'
          enable_pipeline: true
        - addr: 'https://public.dns.iij.jp/dns-query'
          dial_addr: '103.2.57.5:443'
          enable_pipeline: true
        #
        #
        #

        # TW101 DNS 80ms
        - addr: 'https://dns.twnic.tw/dns-query'
          dial_addr: '101.101.101.101:443'
          enable_pipeline: true
        #
        #
        #
        
        # Cisco (OpenDNS) 110ms
        - addr: 'https://doh.opendns.com/dns-query'
          enable_pipeline: true
        #
        #
        #

        # 新加坡DNS https://puredns.org/ 150ms
        - addr: 'https://puredns.org/dns-query'
          enable_pipeline: true
        #  
        # 新加坡DNS https://github.com/bebasid/bebasdns 150ms
        - addr: 'https://dns.bebasid.com/dns-query'
          enable_pipeline: true
        #
        #
        #

        # Quad9 DNS 250ms
        - addr: 'https://9.9.9.9/dns-query'
          dial_addr: '9.9.9.9:443'
          enable_pipeline: true
        - addr: 'https://149.112.112.112/dns-query'
          dial_addr: '149.112.112.112:443'
          enable_pipeline: true
        #
        #
        #
        
        # 备用直连DNS.SB 地址网站 https://dns.sb/doh/        
        # HK 100ms 寄了
        #- addr: 'https://doh.dns.sb/dns-query'
        #- addr: 'https://hk-hkg.doh.sb/dns-query'        
        #  dial_addr: '45.125.0.26:443'
        #  enable_pipeline: true

        # Tokyo 70ms 寄了
        #- addr: 'https://doh.dns.sb/dns-query'
        #  dial_addr: '103.121.210.210:443'
        #  enable_pipeline: true

        # DNS.SB - Osaka, Japan 133ms 丢包严重
        #- addr: 'https://doh.dns.sb/dns-query'
        #  dial_addr: '202.5.221.130:443'
        #  enable_pipeline: true
        #- addr: 'https://doh.dns.sb/dns-query'
        #  dial_addr: '202.5.221.131:443'
        #  enable_pipeline: true
        #- addr: 'https://doh.dns.sb/dns-query'
        #  dial_addr: '202.5.221.132:443'
        #  enable_pipeline: true
        #- addr: 'https://doh.dns.sb/dns-query'
        #  dial_addr: '202.5.221.133:443'
        #  enable_pipeline: true
        #- addr: 'https://doh.dns.sb/dns-query'
        #  dial_addr: '202.5.221.134:443'
        #  enable_pipeline: true

        # seoul 100ms 寄了
        #- addr: 'https://doh.dns.sb/dns-query'
        #  dial_addr: '3.34.32.82:443'
        #  enable_pipeline: true

        # SG 250ms 丢包严重
        #- addr: 'https://doh.dns.sb/dns-query'
        #  dial_addr: '165.22.61.129:443'
        #  enable_pipeline: true

  - tag: Force_forward_local # 强制境内
    type: query_matcher
    args:
      domain:
        - "ext:./geosite.dat:apple-cn"
        - "ext:./geosite.dat:google-cn"
        - "ext:./Force_forward_local.txt"


  - tag: Force_forward_remote # 强制境外
    type: query_matcher
    args:
      domain:
        - "ext:./geosite.dat:gfw"
        - "ext:./geosite.dat:greatfire"
        - "ext:./Force_forward_remote.txt"

  - tag: query_is_local_domain
    type: query_matcher
    args:
      domain:
        - "ext:./geosite.dat:cn"

  - tag: query_is_non_local_domain
    type: query_matcher
    args:
      domain:
        - "ext:./geosite.dat:geolocation-!cn"

  - tag: response_has_local_ip
    type: response_matcher
    args:
      ip:
        - "ext:./geoip.dat:cn"

  - tag: query_is_ad_domain
    type: query_matcher
    args:
      domain:
        - "ext:./geosite.dat:category-ads"
        - "ext:./geosite.dat:category-ads-all"

**Comment by Tmalll** [comment]
> * tag: query_is_ad_domain
>   type: query_matcher
>   args:
>   domain:
>   - "ext:./geosite.dat:category-ads"
>   - "ext:./geosite.dat:category-ads-all"

4.x版本好像 query_matcher 不能直接像3.x那样引用文件了 server跑到大类里面 不知道判定有没有改. 反正4.x的wiki写的乱七八糟,看着人头大....
tag: query_is_ad_domain
type: query_matcher
args:
domain:
- "ext:./geosite.dat:category-ads"
- "ext:./geosite.dat:category-ads-all"

**Comment by liangxyz** [comment]
直接用wiki里的easymosdns

**Comment by Tmalll** [comment]
> 直接用wiki里的easymosdns

还是自己写的判定放心点...

**Comment by hpcex** [comment]
贴个自己的简单的分流的 4.x 配置，供参考。
https://paste2.org/VYvtPFUz

**Comment by hoicau** [comment]
个人的example，供参考
https://github.com/c2xvi/easier-mosdns#%E8%87%AA%E5%8A%A8%E5%88%86%E6%B5%81%E5%B9%BF%E5%91%8A%E5%B1%8F%E8%94%BD%E9%85%8D%E7%BD%AE%E5%80%9F%E5%8A%A9v2ray-rules-dat%E4%B8%AD%E7%9A%84geositedat%E5%92%8Cgeoipdat
